File: seema-node/src/routes/breach.ts
import { Router, Request, Response } from 'express';
import axios, { AxiosError } from 'axios';
import { z } from 'zod';
import prisma from '../lib/prisma.js';
import { authenticate } from '../middleware/auth.js';
import { getTenantFilter } from '../middleware/tenant.js';
import { logAudit } from '../middleware/auditLogger.js';
import logger from '../utils/logger';

const router = Router();

const FASTAPI_URL = process.env.FASTAPI_URL || 'http://api:8000';

// Serialize Prisma's camelCase to the snake_case shape the frontend expects.
function serializeBreach(b: any) {
  // The AI draft is persisted as a JSON string in the DB so we can rehydrate
  // it for the modal without re-calling the AI. Parse defensively here so a
  // malformed legacy row doesn't 500 the whole list endpoint.
  let icoDraft: unknown = null;
  if (b.icoNotificationDraft) {
    try {
      icoDraft = JSON.parse(b.icoNotificationDraft);
    } catch {
      // Treat unparseable values as a plain text draft (older rows or
      // hand-edited DB rows). The modal renders strings as the summary.
      icoDraft = { summary: String(b.icoNotificationDraft) };
    }
  }
  return {
    id: b.id,
    title: b.title,
    description: b.description,
    breach_type: b.breachType,
    severity: b.severity,
    reported_date: b.reportedDate,
    ico_deadline: b.icoDeadline,
    status: b.status,
    notification_status: b.notificationStatus,
    affected_records: b.affectedRecords ?? 0,
    root_cause: b.rootCause ?? null,
    resolution_date: b.resolutionDate ?? null,
    // Task #48 — ICO 72-hour workflow fields.
    ico_notification_draft: icoDraft,
    ico_notification_drafted_at: b.icoNotificationDraftedAt ?? null,
    ico_notified_at: b.icoNotifiedAt ?? null,
    created_at: b.createdAt,
    updated_at: b.updatedAt,
  };
}

// Frontend sends snake_case (breach_type); older callers sent camelCase
// (breachType). Accept both so the route doesn't 400.
const breachSchema = z.object({
  title: z.string().min(1),
  description: z.string().min(1),
  severity: z.string().min(1),
  // snake_case (frontend)
  breach_type: z.string().optional(),
  // camelCase (legacy)
  breachType: z.string().optional(),
}).refine(
  (d) => Boolean(d.breach_type || d.breachType),
  { message: 'breach_type (or breachType) is required' },
);

// GET /compliance/breach-reports
router.get('/compliance/breach-reports', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);

    const reports = await prisma.breachReport.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
    });

    res.json(reports.map(serializeBreach));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch breach reports' });
  }
});

// POST /compliance/breach-report
router.post('/compliance/breach-report', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const parsed = breachSchema.safeParse(req.body);

    if (!parsed.success) {
      res.status(400).json({ error: true, message: 'Validation failed', details: parsed.error.errors });
      return;
    }

    const { title, description, severity } = parsed.data;
    const breachType = (parsed.data.breach_type ?? parsed.data.breachType) as string;
    const reportedDate = new Date();
    const icoDeadline = new Date(reportedDate.getTime() + 72 * 60 * 60 * 1000);

    const breach = await prisma.breachReport.create({
      data: {
        firmId,
        title,
        description,
        breachType,
        severity,
        reportedDate,
        icoDeadline,
        status: 'open',
        notificationStatus: 'pending',
      },
    });

    await prisma.complianceAlert.create({
      data: {
        firmId,
        alertType: 'breach',
        severity,
        title: `Breach Report: ${title}`,
        description: `A new breach report has been filed: ${title}`,
        status: 'open',
      },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'create_breach_report',
      entityType: 'breach_report',
      entityId: breach.id,
      ipAddress: req.ip,
    });

    res.status(201).json(serializeBreach(breach));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to create breach report' });
  }
});

// ---------------------------------------------------------------------------
// Task #48 — ICO 72-hour notification workflow
// ---------------------------------------------------------------------------
// These endpoints sit on the Node API rather than being thin proxies because
// they need to (a) load the Prisma row, (b) call the FastAPI AI service, and
// (c) persist the resulting draft + audit-log it. Doing all three from the
// browser would mean three round-trips and a race window where a draft is
// shown but never saved.

// POST /compliance/breach-reports/:id/draft-ico-notification
// Asks the FastAPI AI service to draft a UK GDPR Article 33 notification for
// this breach, persists the draft + drafted_at on the breach row, and returns
// the structured draft to the caller. Re-calling this endpoint regenerates
// the draft (overwrites the previous one) — the COLP can iterate.
router.post(
  '/compliance/breach-reports/:id/draft-ico-notification',
  authenticate,
  async (req: Request, res: Response) => {
    const { firmId } = getTenantFilter(req);
    const breachId = req.params.id as string;
    try {
      // Verify the breach exists in this firm before round-tripping to AI.
      const breach = await prisma.breachReport.findFirst({
        where: { id: breachId, firmId },
      });
      if (!breach) {
        res.status(404).json({ error: true, message: 'Breach not found' });
        return;
      }

      // Proxy through to the existing FastAPI endpoint, which already
      // contains the BREACH_NOTIFICATION_SYSTEM_PROMPT and citation logic.
      // We forward the JWT so FastAPI's tenant_db_from_jwt dependency can
      // re-derive the firm context.
      const aiResponse = await axios({
        method: 'POST',
        url: `${FASTAPI_URL}/api/ai/draft-ico-notification`,
        data: { breach_id: breachId },
        headers: {
          'Content-Type': 'application/json',
          ...(req.headers.authorization
            ? { Authorization: req.headers.authorization }
            : {}),
          'X-Firm-ID': firmId,
          'X-User-ID': req.user?.userId || '',
          'X-User-Role': req.user?.role || '',
          'X-User-Email': req.user?.email || '',
        },
        // 120s — Claude calls regularly run 30-60s; match aiProxy default.
        timeout: 120000,
      });

      const draft = aiResponse.data;

      // Persist the draft so the modal can reopen it without re-paying for
      // the AI call. Stored as JSON so the structured shape survives.
      const updated = await prisma.breachReport.update({
        where: { id: breachId },
        data: {
          icoNotificationDraft: JSON.stringify(draft),
          icoNotificationDraftedAt: new Date(),
        },
      });

      await logAudit({
        firmId,
        userId: req.user!.userId,
        action: 'draft_ico_notification',
        entityType: 'breach_report',
        entityId: breachId,
        ipAddress: req.ip,
      });

      res.status(200).json({
        draft,
        breach: serializeBreach(updated),
      });
    } catch (err) {
      if (err instanceof AxiosError) {
        const status = err.response?.status || 502;
        const message = err.response?.data?.detail || err.message;
        logger.error('AI draft-ico-notification proxy error', {
          breachId,
          status,
          message,
        });
        res.status(status).json({
          error: true,
          message: status === 502
            ? 'AI service temporarily unavailable — try again in a few seconds'
            : message,
        });
        return;
      }
      logger.error('draft-ico-notification failed', {
        err: err instanceof Error ? err.message : String(err),
        breachId,
      });
      res.status(500).json({ error: true, message: 'Failed to draft ICO notification' });
    }
  },
);

// PATCH /compliance/breach-reports/:id/mark-notified
// Records that the COLP has actually submitted the notification to the ICO.
// Sets ico_notified_at + flips notification_status to 'notified'. Idempotent:
// re-PATCHing does not move the timestamp once set, so we don't lose the
// original notification time if someone clicks twice.
router.patch(
  '/compliance/breach-reports/:id/mark-notified',
  authenticate,
  async (req: Request, res: Response) => {
    const { firmId } = getTenantFilter(req);
    const breachId = req.params.id as string;
    try {
      const breach = await prisma.breachReport.findFirst({
        where: { id: breachId, firmId },
      });
      if (!breach) {
        res.status(404).json({ error: true, message: 'Breach not found' });
        return;
      }

      // Preserve the original notification timestamp on duplicate clicks.
      const notifiedAt = breach.icoNotifiedAt ?? new Date();

      const updated = await prisma.breachReport.update({
        where: { id: breachId },
        data: {
          icoNotifiedAt: notifiedAt,
          notificationStatus: 'notified',
        },
      });

      await logAudit({
        firmId,
        userId: req.user!.userId,
        action: 'mark_breach_ico_notified',
        entityType: 'breach_report',
        entityId: breachId,
        ipAddress: req.ip,
      });

      res.status(200).json(serializeBreach(updated));
    } catch (err) {
      logger.error('mark-notified failed', {
        err: err instanceof Error ? err.message : String(err),
        breachId,
      });
      res.status(500).json({ error: true, message: 'Failed to mark breach as notified' });
    }
  },
);

export default router;

File: seema-api/routers/breach.py
"""Breach reporting router for GDPR and data protection compliance."""
import uuid
import logging
from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from middleware.tenant_rls import tenant_db_from_jwt, bypass_db
from middleware.auth import get_current_user, CurrentUser
from services.audit_logger import log_audit
from models.breach import BreachReport
from models.compliance import ComplianceAlert

logger = logging.getLogger("seema.breach")

router = APIRouter()

class CreateBreachReportRequest(BaseModel):
    title: str
    description: str | None = None
    breach_type: str | None = None
    severity: str = "medium"
    reported_date: datetime | None = None
    ico_deadline: datetime | None = None
    affected_records: int = 0
    root_cause: str | None = None

@router.get("/compliance/breach-reports")
async def get_breach_reports(
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """List breach reports."""
    result = await db.execute(
        select(BreachReport).where(
            BreachReport.firm_id == user.firm_id
        ).order_by(BreachReport.created_at.desc())
    )
    breaches = result.scalars().all()

    return [
        {
            "id": b.id,
            "title": b.title,
            "description": b.description,
            "breach_type": b.breach_type,
            "severity": b.severity,
            "status": b.status,
            "reported_date": str(b.reported_date) if b.reported_date else None,
            "ico_deadline": str(b.ico_deadline) if b.ico_deadline else None,
            "notification_status": b.notification_status,
            "affected_records": b.affected_records,
            "root_cause": b.root_cause,
            "resolution_date": str(b.resolution_date) if b.resolution_date else None,
            "remediation_plan_id": b.remediation_plan_id,
            "created_at": str(b.created_at),
            "updated_at": str(b.updated_at),
        }
        for b in breaches
    ]

@router.post("/compliance/breach-report")
async def create_breach_report(
    req: CreateBreachReportRequest,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Create breach report with auto-calculated ICO deadline and alert generation."""
    reported = req.reported_date or datetime.now(timezone.utc)
    # Auto-calculate ICO 72-hour deadline if not provided
    ico_deadline = req.ico_deadline or (reported + timedelta(hours=72))

    breach_id = str(uuid.uuid4())
    breach = BreachReport(
        id=breach_id,
        firm_id=user.firm_id,
        title=req.title,
        description=req.description,
        breach_type=req.breach_type,
        severity=req.severity,
        status="open",
        reported_date=reported,
        ico_deadline=ico_deadline,
        affected_records=req.affected_records,
        root_cause=req.root_cause,
        notification_status="pending",
    )
    db.add(breach)

    # Auto-create a CRITICAL compliance alert for the COLP
    alert_severity = "critical" if req.severity in ("high", "critical") else "high"
    alert = ComplianceAlert(
        id=str(uuid.uuid4()),
        firm_id=user.firm_id,
        alert_type="breach_reported",
        severity=alert_severity,
        title=f"Breach Reported: {req.title}",
        description=(
            f"A {req.breach_type or 'data'} breach has been reported. "
            f"Severity: {req.severity}. "
            f"ICO 72-hour notification deadline: {ico_deadline.strftime('%d %B %Y %H:%M UTC')}. "
            f"Immediate assessment required."
        ),
        action_required=(
            "1. Assess whether ICO notification is required under UK GDPR Article 33. "
            "2. If notifiable, submit via ICO breach reporting tool within 72 hours. "
            "3. Assess whether affected individuals must be notified under Article 34. "
            "4. Document all decisions in Seema."
        ),
        status="open",
    )
    db.add(alert)

    await db.flush()

    await log_audit(
        db=db,
        firm_id=user.firm_id,
        action="created",
        entity_type="breach_report",
        entity_id=breach.id,
        user_id=user.user_id,
        details=f"Breach report created: {req.title} - Severity: {req.severity}. ICO deadline: {ico_deadline.isoformat()}",
    )

    # Fire breach alert email asynchronously (best-effort, don't block response)
    try:
        from models.firm import Firm
        firm_result = await db.execute(select(Firm).where(Firm.id == user.firm_id))
        firm = firm_result.scalar_one_or_none()
        if firm and firm.email:
            from services.email_service import EmailService
            email_svc = EmailService()
            email_svc.send_breach_alert(
                to_email=firm.email,
                to_name=firm.colp_name or "COLP",
                breach_title=req.title,
                breach_category=req.breach_type or "data",
                reported_at=reported.strftime("%d %B %Y %H:%M UTC"),
                ico_deadline=ico_deadline.strftime("%d %B %Y %H:%M UTC"),
                firm_name=firm.name,
            )
            logger.info(f"Breach alert email sent to COLP ({firm.email}) for breach {breach_id}")
    except Exception as e:
        # Email failure should not block breach creation
        logger.warning(f"Failed to send breach alert email: {e}")

    return {
        "id": breach.id,
        "title": breach.title,
        "severity": breach.severity,
        "status": breach.status,
        "ico_deadline": str(ico_deadline),
        "alert_id": alert.id,
        "created_at": str(breach.created_at),
    }
